### PR TITLE
Add IPPool CRD  

### PR DESCRIPTION
**Pull request by zhengxiexie**

### CODE REVIEW

File: pkg/apis/v1alpha1/ippool_types.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+type IPPoolList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []IPPool `json:"items"`
+}
+
+// IPPoolSpec defines the desired state of IPPool.
+type IPPoolSpec struct {
+	// Type defines the type of this IPPool, public or private.
+	Type string `json:"type"`
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/ippool_types.go:36]: We can add // +kubebuilder:validation:Enum=public;private

File: pkg/apis/v1alpha1/ippool_types.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/* Copyright © 2021 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/ippool_types.go:2]: change 2021 to 2022.

File: pkg/apis/v1alpha1/ippool_types.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+// SubnetRequest defines the subnet allocation request.
+type SubnetRequest struct {
+	// PrefixLength defines prefix length for this subnet.
+	// +optional
+	PrefixLength int `json:"prefixLength,omitempty"`
+
+	// IPFamily defines the IP family type for this subnet, could be IPv4 or IPv6.
+	// This is optional, the default is IPv4.
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/ippool_types.go:56]: // +kubebuilder:default=IPv4 to define the default value?

File: go.sum [hunk 4]
```diff
@@ -87,6 +87,7 @@ github.com/blang/semver v3.5.1+incompatible/go.mod h1:kRBLl5iJ+tD4TcOOxsy/0fnweb
 github.com/census-instrumentation/opencensus-proto v0.2.1/go.mod h1:f6KPmirojxKA12rnyqOA5BBL4O983OfeGPqjHWSTneU=
 github.com/certifi/gocertifi v0.0.0-20191021191039-0944d244cd40/go.mod h1:sGbDF6GwGcLpkNXPUTkMRoywsNa/ol15pxFe6ERfguA=
 github.com/certifi/gocertifi v0.0.0-20200922220541-2c3bb06c6054/go.mod h1:sGbDF6GwGcLpkNXPUTkMRoywsNa/ol15pxFe6ERfguA=
+github.com/cespare/xxhash v1.1.0 h1:a6HrQnmkObjyL+Gs60czilIUGqrzKutQD6XZog3p+ko=
```
Comment (lxiaopei) [inline go.sum:90]: shall we add changes for go.mod and go.sum when adding controller?

File: build/yaml/crd/nsx.vmware.com_ippools.yaml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                      type: integer
+                  required:
+                  - name
+                  type: object
+                type: array
+              type:
+                description: Type defines the type of this IPPool, public or private.
+                type: string
+            required:
+            - subnets
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_ippools.yaml:62]: subnets and type should be optional, right? @heypnus

File: pkg/apis/v1alpha1/ippool_types.go (hunk 2 above)
Comment (zhengxiexie) [reply pkg/apis/v1alpha1/ippool_types.go:2]: ok

File: pkg/apis/v1alpha1/ippool_types.go (hunk 1 above)
Comment (zhengxiexie) [reply pkg/apis/v1alpha1/ippool_types.go:36]: ok

File: pkg/apis/v1alpha1/ippool_types.go (hunk 3 above)
Comment (zhengxiexie) [reply pkg/apis/v1alpha1/ippool_types.go:56]: ok

File: go.sum (hunk 4 above)
Comment (zhengxiexie) [reply go.sum:90]: ok

File: build/yaml/crd/nsx.vmware.com_ippools.yaml (hunk 5 above)
Comment (zhengxiexie) [reply build/yaml/crd/nsx.vmware.com_ippools.yaml:62]: ok

File: build/yaml/crd/nsx.vmware.com_ippools.yaml [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            type: object
+          spec:
+            description: IPPoolSpec defines the desired state of IPPool.
+            properties:
+              subnets:
+                description: Subnets defines set of subnets need to be allocated.
+                items:
+                  description: SubnetRequest defines the subnet allocation request.
+                  properties:
+                    ipFamily:
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_ippools.yaml:44]: add enum

File: build/yaml/crd/nsx.vmware.com_ippools.yaml [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+spec:
+  group: nsx.vmware.com
+  names:
+    kind: IPPool
+    listKind: IPPoolList
+    plural: ippools
+    singular: ippool
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_ippools.yaml:19]: `v1alpha1`: for CRD upgrade from existing IPPool CRD, do we need to update the version here?

File: build/yaml/crd/nsx.vmware.com_ippools.yaml (hunk 6 above)
Comment (zhengxiexie) [reply build/yaml/crd/nsx.vmware.com_ippools.yaml:44]: ok

File: build/yaml/crd/nsx.vmware.com_ippools.yaml (hunk 7 above)
Comment (zhengxiexie) [reply build/yaml/crd/nsx.vmware.com_ippools.yaml:19]: v1alpha2?

File: PROJECT [hunk 8]
```diff
@@ -20,4 +20,12 @@ resources:
   kind: SecurityPolicy
   path: github.com/vmware-tanzu/nsx-operator/pkg/api/v1alpha1
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
```
Comment (heypnus) [inline PROJECT:26]: If you don't add the controller in this patch, `controller: true` shouldn't exist.

Comment (zhengxiexie) [reply PROJECT:26]: ok

**Pull request closed without merging**